libraries/chain/controller.cpp

```diff
@@ -1962,7 +1962,9 @@ struct controller_impl {
             } // end if exception
          } /// end for each block in branch
 
-         unapplied_branch = std::move( branches.second );
+         if ( read_mode == db_read_mode::SPECULATIVE ) {
+            unapplied_branch = std::move( branches.second );
+         }
```

Comment:
This is not ideal in that we add forked branches after applying the new branch which might have many of the same transactions.